File: cloud-run-function-kotlin/app/src/main/kotlin/org/phdezann/energy/service/PowerConverter.kt
package org.phdezann.energy.service

import org.phdezann.energy.formater.TimeFormatter
import org.phdezann.energy.json.PowerEntry
import org.phdezann.energy.model.PowerMatch
import org.phdezann.energy.support.FileUtils

class PowerConverter {

    private val powerEntries: Array<PowerEntry>
    private val timeFormatter = TimeFormatter()

    init {
        val json = FileUtils.read("power.json")
        powerEntries = JsonSerializer().readValue(json, Array<PowerEntry>::class.java)
    }

    fun convert(milliWattHour: Double): List<PowerMatch> {
        val matches = mutableListOf<PowerMatch>()
        val wattHour = milliWattHour / 1000

        for (entry in powerEntries) {
            val hours = wattHour / entry.powerConsumptionWatt
            val msg = timeFormatter.format(hours)
            matches.add(PowerMatch(entry.label, msg))
        }

        return matches
    }

}

File: src/energy-converter/power-converter.ts
import powerData, { PowerEntry } from './data/power';
import TimeFormatter from './time-formatter';

export interface PowerMatch {
  label: string;
  message: string;
}

class PowerConverter {
  private readonly powerEntries: PowerEntry[];
  private timeFormatter: TimeFormatter;

  constructor() {
    this.powerEntries = powerData;
    this.timeFormatter = new TimeFormatter();
  }

  convert(milliWattHour: number): PowerMatch[] {
    const matches: PowerMatch[] = [];
    const wattHour = milliWattHour / 1000;

    for (const entry of this.powerEntries) {
      const hours = wattHour / entry.power_consumption_watt;
      const msg = this.timeFormatter.format(hours);
      matches.push({ label: entry.label, message: msg });
    }

    return matches;
  }
}

export default PowerConverter;
